feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go

```diff
@@ -0,0 +1,625 @@
+			EthernetType: &gdpEtherType,
+		},
+		IngressPort: fmt.Sprint(portId),
+	}
+}
+
+// GetTableEntry creates wbb acl entry related to GDP.
+func (gdp *GDPPacketIO) GetTableEntry(delete bool) []*p4rtutils.ACLWbbIngressTableEntryInfo {
+	actionType := p4_v1.Update_INSERT
+	if delete {
+		actionType = p4_v1.Update_DELETE
+	}
+	return []*p4rtutils.ACLWbbIngressTableEntryInfo{{
+		Type:          actionType,
+		EtherType:     0x6007,
+		EtherTypeMask: 0xFFFF,
+		Priority:      1,
+	}}
+}
+
+// writeTableEntry programs or deletes p4rt table entry
+func writeTableEntry(args *testArgs, t *testing.T, packetIO PacketIO, delete bool) error {
+	t.Helper()
+
+	p4Info, err := utils.P4InfoLoad(p4InfoFile)
+	if err != nil {
+		return errors.New("Errors seen when loading p4info file.")
+	}
+
+	if err := args.handle.SetForwardingPipelineConfig(&p4_v1.SetForwardingPipelineConfigRequest{
+		DeviceId:   args.deviceID,
+		ElectionId: &p4_v1.Uint128{High: args.highID, Low: args.lowID},
+		Action:     p4_v1.SetForwardingPipelineConfigRequest_VERIFY_AND_COMMIT,
+		Config: &p4_v1.ForwardingPipelineConfig{
+			P4Info: &p4Info,
+			Cookie: &p4_v1.ForwardingPipelineConfig_Cookie{
+				Cookie: 159,
+			},
+		},
+	}); err != nil {
+		return errors.New("Errors seen when sending SetForwardingPipelineConfig.")
+	}
+
+	if err := args.handle.Write(&p4_v1.WriteRequest{
+		DeviceId:   args.deviceID,
+		ElectionId: &p4_v1.Uint128{High: args.highID, Low: args.lowID},
+		Updates: p4rtutils.ACLWbbIngressTableEntryGet(
+			packetIO.GetTableEntry(delete),
+		),
+		Atomicity: p4_v1.WriteRequest_CONTINUE_ON_ERROR,
+	}); err != nil {
+		t.Logf("Write Error: %v", err)
+		return err
+	}
+	return nil
+}
+
+// Function to read a table entry returns response/error
+func readTableEntry(args *testArgs, t *testing.T) (*p4_v1.ReadResponse, error) {
+	t.Helper()
+	readMsg := &p4_v1.ReadRequest{
+		DeviceId: args.deviceID,
+		Entities: []*p4_v1.Entity{
+			{
+				Entity: &p4_v1.Entity_TableEntry{
+					TableEntry: &p4_v1.TableEntry{
+						TableId: p4rtutils.WbbTableMap["acl_wbb_ingress_table"],
+					},
+				},
+			},
+		},
+	}
+	readClient, err := args.handle.Read(readMsg)
+	if err != nil {
+		t.Logf("ReadClient Error: %v", err)
+		return nil, err
+	}
+	readResp, err := readClient.Recv()
+	if err != nil {
+		t.Logf("readResp Error: %v", err)
+		return nil, err
+	}
+	return readResp, nil
+}
+
+// Every client/controller should be able to read
+func canRead(t *testing.T, args *testArgs) (bool, error) {
+	_, readErr := readTableEntry(args, t)
+	if readErr != nil {
+		return false, readErr
+	}
+	return true, nil
+}
+
+// Only Primary should be able to write
+func canWrite(t *testing.T, args *testArgs) (bool, error) {
+	pktIO := getGDPParameter(t)
+	writeErr := writeTableEntry(args, t, pktIO, false)
+	if writeErr != nil {
+		if args.wantWrite == false {
+			t.Logf("Expected P4Client error: Client write permission (highID %d, lowID %d): want %v, got %v", args.highID, args.lowID, args.wantWrite, false)
```

Comment:
Logic error here. Should it not be args.wantWrite==true?

Reply:
We're only hitting this block when the write actually fails (err != nil) and if args.wantWrite is false, then it's an expected error, hence the t.Logf